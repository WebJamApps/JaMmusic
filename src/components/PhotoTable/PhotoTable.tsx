import React, { Dispatch } from 'react';
import MUIDataTable, { MUIDataTableColumnDef } from 'mui-datatables';
import ReactHtmlParser from 'react-html-parser';
import type { AGClientSocket } from 'socketcluster-client';
import 'core-js/stable';
import { connect } from 'react-redux';
import superagent from 'superagent';
import type { AnyAction } from 'redux';
import mapStoreToProps, { Iimage } from '../../redux/mapStoreToProps';

type Pprops = {
  dispatch: Dispatch<AnyAction>,
  picUpdated?: boolean,
  auth: { token: string },
  images: Iimage[],
  scc?: AGClientSocket
};
interface Pstate {
  columns: MUIDataTableColumnDef[]
}
export class PhotoTable extends React.Component<Pprops, Pstate> {
  superagent: superagent.SuperAgentStatic;

  constructor(props: Readonly<Pprops>) {
    super(props);
    this.superagent = superagent;
    this.setColumns = this.setColumns.bind(this);
    this.handleHideTable = this.handleHideTable.bind(this);
    this.addThumbs = this.addThumbs.bind(this);
    this.state = {
      columns: [],
    };
  }

  componentDidMount(): void { this.setColumns(); }

  handleHideTable(): boolean {
    const { dispatch } = this.props;
    dispatch({ type: 'SHOW_TABLE', showTable: false });
    return true;
  }

  setColumns(): void {
    const columns: MUIDataTableColumnDef[] = [];
    const titles = ['Thumbnail', 'Title', 'Caption', 'Link', 'Type', 'Modify'];
    for (let i = 0; i < titles.length; i += 1) {
      const label = titles[i];// eslint-disable-line security/detect-object-injection
      columns.push({
        name: label.toLowerCase(),
        label,
        options: {
          filter: false,
          sort: true,
          customBodyRender: (value: string) => (
            <div style={{
              margin: 0, fontSize: '12pt', maxWidth: '200px',
            }}
            >
              {label !== 'Modify' ? ReactHtmlParser(value) : value}
            </div>
          ),
        },
      });
    }
    this.setState({ columns });
  }

  /* let res: superagent.Response;
      try {
        res = await this.superagent.delete(`${process.env.BackendUrl}/book/${id}`)
          .set('Authorization', `Bearer ${auth.token}`).set('Accept', 'application/json');
      } catch (e) { return `${e.message}`; }
      if (res.status === 200) { window.location.reload(); return 'deleted pic'; }
      return `${res.status} ${res.body}`;
    }
    return 'no delete'; */

  deletePic(id: string | undefined): boolean { // eslint-disable-next-line no-restricted-globals
    const result = confirm('Deleting picture, are you sure?');// eslint-disable-line no-alert
    if (result) {
      const { scc, auth } = this.props;
      const pic = { id };
      if (scc && auth) {
        scc.transmit('deleteImage', { pic, token: auth.token });
        window.location.assign('/music');
        return true;
      } return false;
    } return false;
  }

  editPic(editImage: Iimage): boolean {
    const { dispatch } = this.props;
    // eslint-disable-next-line no-param-reassign
    // delete picData.modify;
    dispatch({ type: 'EDIT_PIC', editImage });
    return true;
  }

  addThumbs(arr: Iimage[]): Iimage[] {
    const newArr = arr;/* eslint-disable security/detect-object-injection */
    for (let i = 0; i < arr.length; i += 1) { // eslint-disable-next-line security/detect-object-injection
      newArr[i].thumbnail = `<img src=${arr[i].url} width="200px"/>`;
      const deletePicId = `deletePic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
      const editPicId = `editPic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
      newArr[i].link = `<a href=${arr[i].url} target="_blank">click to view</a>`;// eslint-disable-line security/detect-object-injection
      newArr[i].caption = newArr[i].comments === 'showCaption' ? 'display' : 'hide';
      newArr[i].modify = (// eslint-disable-line security/detect-object-injection
        <div>
          <button type="button" id={deletePicId} onClick={() => this.deletePic(newArr[i]._id)}>Delete Pic</button>
          <p>{' '}</p>
          <button type="button" id={editPicId} onClick={() => { this.editPic(newArr[i]); this.handleHideTable(); }}>
            Edit Pic
          </button>
        </div>
      );
    }
    return newArr;
  }

  render(): JSX.Element {
    const { columns } = this.state;
    const {
      images,
    } = this.props;
    let arr: Iimage[] = images;
    arr = this.addThumbs(arr);
    return (
      <div className="photoTable">
        <div style={{ maxWidth: '9in', margin: 'auto' }}>
          <MUIDataTable
            options={{
              filterType: 'dropdown',
              pagination: false,
              responsive: 'standard',
              filter: false,
              download: false,
              search: false,
              print: false,
              viewColumns: false,
              selectableRows: 'none',
              fixedHeader: false,
            }}
            columns={columns}
            data={arr}
            title=""
          />
        </div>
      </div>
    );
  }
}

export default connect(mapStoreToProps, null)(PhotoTable);
