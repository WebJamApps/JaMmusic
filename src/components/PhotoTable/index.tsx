import type { Dispatch } from 'react';
import MUIDataTable from 'mui-datatables';
import type { AnyAction } from 'redux';
import type { Iimage } from '../../redux/mapStoreToProps';
import utils from './photoTable.utils';

export const PhotoTable = (props:any):JSX.Element => {

  let dataArr: Iimage[] = props.images;
  dataArr = utils.addThumbs(dataArr, props.controller, props.dispatch);
  return (
      <div className="photoTable">
        <div style={{ maxWidth: '9in', margin: 'auto' }}>
          <MUIDataTable
            options={{
              filterType: 'dropdown',
              pagination: false,
              responsive: 'standard',
              filter: false,
              download: false,
              search: false,
              print: false,
              viewColumns: false,
              selectableRows: 'none',
              fixedHeader: false,
            }}
            columns={utils.setColumns()}
            data={dataArr}
            title=""
          />
        </div>
      </div>
  );
};
