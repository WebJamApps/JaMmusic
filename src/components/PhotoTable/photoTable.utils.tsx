import type { MUIDataTableColumnDef } from 'mui-datatables';
import HtmlReactParser from 'html-react-parser';

import type { Iimage } from '../../redux/mapStoreToProps';

const editPic = (data: Iimage, dispatch: (arg0: { type: string; data: Iimage; }) => void): boolean => {
  delete data.modify;
  dispatch({ type: 'EDIT_PIC', data });
  return true;
};

const addThumbs = (
  arr: Iimage[],
  controller: { deleteData: (arg0: string, arg1: string | undefined) => void; },
  dispatch: (arg0: { type: string; data: Iimage; }) => void,
): Iimage[] => {
  const newArr = arr;/* eslint-disable security/detect-object-injection */
  for (let i = 0; i < arr.length; i += 1) { // eslint-disable-next-line security/detect-object-injection
    newArr[i].thumbnail = `<img src=${arr[i].url} width="200px"/>`;
    const deletePicId = `deletePic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
    const editPicId = `editPic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
    newArr[i].link = `<a href=${arr[i].url} target="_blank">click to view</a>`;// eslint-disable-line security/detect-object-injection
    newArr[i].caption = newArr[i].comments === 'showCaption' ? 'display' : 'hide';
    newArr[i].modify = (// eslint-disable-line security/detect-object-injection
      <div>
        <button
          type="button"
          id={deletePicId}
          onClick={() => {
            controller.deleteData('deleteImage', newArr[i]._id);
          }}
        >
          Delete Pic
        </button>
        <p>{' '}</p>
        <button type="button" id={editPicId} onClick={() => editPic(newArr[i], dispatch)}>
          Edit Pic
        </button>
      </div>
    );
  }
  return newArr;
};

const setCustomBody = (value:any, label:string):JSX.Element => {
  if (typeof value !== 'string' && label !== 'Modify') value = '';
  return (
    <div style={{
      margin: 0, fontSize: '12pt', maxWidth: '200px',
    }}
    >
      {label !== 'Modify' ? HtmlReactParser(value) : value}
    </div>
  );
};

const setColumns = (): MUIDataTableColumnDef[] => {
  const columns: MUIDataTableColumnDef[] = [];
  const titles = ['Thumbnail', 'Title', 'Caption', 'Link', 'Type', 'Modify'];
  for (let i = 0; i < titles.length; i += 1) {
    const label = titles[i];// eslint-disable-line security/detect-object-injection
    columns.push({
      name: label.toLowerCase(),
      label,
      options: {
        filter: false,
        sort: true,
        customBodyRender: (value:any) => setCustomBody(value, label),
      },
    });
  }
  return columns;
};

export default {
  editPic, addThumbs, setColumns, setCustomBody,
};
