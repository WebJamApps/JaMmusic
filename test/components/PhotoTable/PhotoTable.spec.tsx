/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { shallow, ShallowWrapper, HTMLAttributes } from 'enzyme';
import { PhotoTable } from '../../../src/components/PhotoTable/PhotoTable';

describe('PhotoTable', () => {
  let props: { auth: any; images: any; },
    wrapper: any, r: ShallowWrapper<HTMLAttributes, any, React.Component<any, any, any>>;
  const controller:any = { deleteData: jest.fn() };
  beforeEach(() => {
    props = {
      auth: { token: 'token' },
      images: [{
        _id: '456', url: 'url', title: 'title', type: 'youthPics', comments: 'showCaption',
      }],
    };
    wrapper = shallow<PhotoTable>(<PhotoTable
      auth={props.auth}
      images={props.images}
      dispatch={(fun) => fun}
      controller={controller}
    />);
  });
  it('renders correctly', () => { expect(wrapper).toMatchSnapshot(); });
  it('sets the columns with customBodyRender', () => {
    expect(typeof wrapper.instance().setColumns).toBe('function');
    wrapper.instance().setColumns();
    const custom = wrapper.instance().state.columns[0].options.customBodyRender('<a href="http://collegelutheran.org/"'
      + ' rel="noopener noreferrer" target="_blank">College Lutheran Church</a>');
    expect(custom.type).toBe('div');
  });
  it('sets the columns with customBodyRender and no data', () => {
    expect(typeof wrapper.instance().setColumns).toBe('function');
    wrapper.instance().setColumns();
    const custom = wrapper.instance().state.columns[0].options.customBodyRender(null);
    expect(custom.type).toBe('div');
  });
  it('sets the columns with customBodyRender for Modify column', () => {
    const buttonjsx = (<button type="button" style={{ display: 'block' }}>howdy</button>);
    expect(typeof wrapper.instance().setColumns).toBe('function');
    wrapper.instance().setColumns();
    const custom = wrapper.instance().state.columns[5].options.customBodyRender(buttonjsx);
    expect(custom.type).toBe('div');
  });
  it('handles click on delete pic button', () => {
    wrapper.update();
    const newArr = wrapper.instance().addThumbs([{ url: 'url', _id: '456' }]);
    const button = shallow(newArr[0].modify);
    button.find('button#deletePic456').simulate('click');
    expect(controller.deleteData).toHaveBeenCalled();
  });
  it('handles click on edit pic button', () => {
    wrapper.instance().editPic = jest.fn();
    wrapper.update();
    const newArr = wrapper.instance().addThumbs([{ url: 'url', _id: '456' }]);
    const button = shallow(newArr[0].modify);
    r = button.find('button#editPic456').simulate('click');
    expect(wrapper.instance().editPic).toHaveBeenCalled();
  });
  it('stores the edit pic data to redux', () => {
    r = wrapper.instance().editPic({});
    expect(r).toBe(true);
  });
});
