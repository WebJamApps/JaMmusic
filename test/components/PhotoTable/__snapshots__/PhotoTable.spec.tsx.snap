// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoTable renders correctly 1`] = `ShallowWrapper {}`;
